# Dna-Sequencing-and-Matching
To create a genetic matching and detection of exact matches  between two given DNA samples and finding an approximate match  for a given DNA sample. 
# Overview:
This project is a genetic analysis application that performs exact and approximate DNA sequence matching, mutation detection, and builds phylogenetic trees to show evolutionary relationships.

Technologies and Algorithms Used:

- Smith-Waterman Algorithm (for local alignment)

- Boyer-Moore Algorithm (for approximate pattern matching)

- Hamming Distance (for mutation detection)

- UPGMA (for phylogenetic tree construction)

Tkinter GUI

NetworkX and matplotlib for graph visualization

# Key Features:

- Matches two DNA sequences for exact and approximate results

- Detects point mutations, useful in early cancer detection

- Creates evolutionary trees based on genetic distances

- Visualizes sequences as weighted graphs

- GUI for ease of interaction

# Conclusion:
A complete bioinformatics toolkit that can assist researchers in genetic analysis, mutation identification, and species comparison using graphical and algorithmic methods.

# Note:
download all the python files in same folder alonside with fasta data folder in the same folder as other python files and then execute the final code python file.

